src/s3/client.rs

```diff
@@ -203,6 +204,91 @@ fn parse_list_objects_common_prefixes(
     Ok(())
 }
 
+/// Client Builder manufactures a Client using given parameters.
+#[derive(Debug, Default)]
+pub struct ClientBuilder {
+    base_url: BaseUrl,
+    provider: Option<Arc<Box<(dyn Provider + Send + Sync + 'static)>>>,
+    ssl_cert_file: Option<PathBuf>,
+    ignore_cert_check: Option<bool>,
+    user_agent: Option<String>,
+}
+
+impl ClientBuilder {
+    /// Creates a builder given a base URL for the MinIO service or other AWS S3
+    /// compatible object storage service.
+    pub fn new(base_url: BaseUrl) -> Self {
+        let mut c = ClientBuilder::default();
+        c.base_url = base_url;
+        c
+    }
+
+    /// Set the credential provider. If not set anonymous access is used.
+    pub fn provider(
+        mut self,
+        provider: Option<Box<(dyn Provider + Send + Sync + 'static)>>,
+    ) -> Self {
+        self.provider = provider.map(Arc::new);
+        self
+    }
+
+    /// Set the user agent.
+    pub fn user_agent(mut self, user_agent: Option<String>) -> Self {
+        self.user_agent = user_agent;
+        self
+    }
+
+    /// Set file for loading a trust certificate.
+    pub fn ssl_cert_file(mut self, ssl_cert_file: Option<&Path>) -> Self {
+        self.ssl_cert_file = ssl_cert_file.map(PathBuf::from);
+        self
+    }
+
+    /// Set the ignore_cert_check.
+    pub fn ignore_cert_check(mut self, ignore_cert_check: Option<bool>) -> Self {
+        self.ignore_cert_check = ignore_cert_check;
+        self
+    }
+
+    /// Build the Client.
+    pub fn build(self) -> Result<Client, Error> {
+        let mut builder = reqwest::Client::builder().no_gzip();
+
+        if let Some(v) = self.user_agent {
+            builder = builder.user_agent(v);
```

Comment:
In user_agent, SDK specific information must be present. Users are allowed to add additional app_name/app_version like https://github.com/minio/minio-py/blob/master/minio/api.py#L493. We should not allow to override entire value.

Reply:
Done